torch/_refs/__init__.py

```diff
@@ -3284,12 +3308,45 @@ def new_zeros(
     layout: Optional[torch.layout] = None,
     device: Optional[torch.device] = None,
     pin_memory: bool = False,
+    requires_grad: bool = False,
 ) -> TensorLikeType:
-    r = a.new_empty(
-        size, dtype=dtype, layout=layout, device=device, pin_memory=pin_memory
+    dtype = a.dtype if dtype is None else dtype
+    layout = a.layout if layout is None else layout
+    device = a.device if device is None else device
+
+    return torch.full(
+        size,
+        False,
+        dtype=dtype,
+        layout=layout,
+        device=device,
+        pin_memory=pin_memory,
+        requires_grad=requires_grad,
+    )
```

Comment:
Alas, I don't think we can do this for the `new_*` functions. These functions have the property that the returned tensor is the same class as the input tensor if the input is not a `Tensor` but a subclass of it.

Now, currently, `new_empty` does not have this property either, as it dispatches to `torch.empty`. Is this an issue @ngimel @mruberry?

Reply:
Interesting! I wonder if we're testing for that (if not, we should be!).

We should probably note this as an issue for now.

Reply:
> I wonder if we're testing for that (if not, we should be!).

We're not, but we definitely should. There's an issue for it over at https://github.com/pytorch/pytorch/issues/82239 .

I talked a bit with @SherlockNoMad about this offline: it's generally OK to decompose `new_zeros` into `zeros` inside the ref of an out-of-place operator, as long as the ref only contains out-of-place operations. We forgot about the case where the user is just directly calling `new_zeros` though, so this is a good catch, @Lezcano.

It seems like we want to tag our refs/decompositions with something that says when they are allowed to run. If we are running solely on regular Tensors, then many decompositions (like the new_zeros -> full) apply; if we are running on Tensor subclasses, then we cannot actually decompose a user's new_zeros call into full.

Reply:
Probably, the correct way of doing this would be to dispatch all these to either `new_full`  and implement that one correctly. I think we should try to avoid yet another way to split how a decomposition works. We are already going to have many branches (decompose at a Python level or not, decompose at an ATen level or not, decompose for backward or not, etc).

Reply:
I think we should be willing to write decomps that are not subclass preserving, and just have them marked as such so they are not used in situations where you have a tensor subclass that is not a mode and you need to prop the subclass. The whole composite compliance new stuff is from the era when we didn't have modes; with modes this sort of thing is a lot less pressing